source/validate_ext_inst.cpp

```diff
@@ -0,0 +1,755 @@
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected operand X type to be equal to Result Type";
+        }
+
+        if (y_type != result_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected operand Y type to be equal to Result Type";
+        }
+        break;
+      }
+
+      case GLSLstd450Refract: {
+        if (!_.IsFloatScalarOrVectorType(result_type)) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected Result Type to be a float scalar or vector type";
+        }
+
+        const uint32_t i_type = _.GetOperandTypeId(inst, 4);
+        const uint32_t n_type = _.GetOperandTypeId(inst, 5);
+        const uint32_t eta_type = _.GetOperandTypeId(inst, 6);
+
+        if (result_type != i_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected operand I to be of type equal to Result Type";
+        }
+
+        if (result_type != n_type) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected operand N to be of type equal to Result Type";
+        }
+
+        const uint32_t eta_type_bit_width = _.GetBitWidth(eta_type);
+        if (!_.IsFloatScalarType(eta_type) ||
+            (eta_type_bit_width != 16 && eta_type_bit_width != 32)) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected operand Eta to be a 16 or 32-bit float scalar";
+        }
+        break;
+      }
+
+      case GLSLstd450InterpolateAtCentroid:
+      case GLSLstd450InterpolateAtSample:
+      case GLSLstd450InterpolateAtOffset: {
+        if (!_.HasCapability(SpvCapabilityInterpolationFunction)) {
+          return _.diag(SPV_ERROR_INVALID_CAPABILITY)
+                 << ext_inst_name()
+                 << " requires capability InterpolationFunction";
+        }
+
+        if (!_.IsFloatScalarOrVectorType(result_type) ||
+            _.GetBitWidth(result_type) != 32) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected Result Type to be a 32-bit float scalar "
+                 << "or vector type";
+        }
+
+        const uint32_t interpolant_type = _.GetOperandTypeId(inst, 4);
+        uint32_t interpolant_storage_class = 0;
+        uint32_t interpolant_data_type = 0;
+        if (!_.GetPointerTypeInfo(interpolant_type, &interpolant_data_type,
+                                  &interpolant_storage_class)) {
+          return _.diag(SPV_ERROR_INVALID_DATA)
+                 << ext_inst_name() << ": "
+                 << "expected Interpolant to be a pointer";
+        }
+
+        if (result_type != interpolant_data_type) {
```

Comment:
Yes, thanks for thinking this one through.  I've filed Khronos internal issue 260 to fix the description in the spec.